// kefir.react.native/lib/combine.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.asProperty = undefined;

var _kefir = require("kefir");

var Kefir = _interopRequireWildcard(_kefir);

var _ramda = require("ramda");

var R = _interopRequireWildcard(_ramda);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function forEach(template, fn) {
  if (template instanceof Kefir.Observable) {
    fn(template);
  } else {
    var _constructor = template && template.constructor;

    if (_constructor === Array) for (var i = 0, n = template.length; i < n; ++i) {
      forEach(template[i], fn);
    } else if (_constructor === Object) for (var k in template) {
      forEach(template[k], fn);
    }
  }
}

function count(template) {
  var count = 0;
  forEach(template, function () {
    return count += 1;
  });
  return count;
}

function subscribe(template, handlers, self) {
  var index = -1;
  forEach(template, function (observable) {
    var i = ++index;
    var handler = function handler(e) {
      return self._handleAny(i, e);
    };
    handlers[i] = handler;
    observable.onAny(handler);
  });
}

function unsubscribe(template, handlers) {
  var index = -1;
  forEach(template, function (observable) {
    var handler = handlers[++index];
    if (handler) observable.offAny(handler);
  });
}

function combine(template, values, state) {
  if (template instanceof Kefir.Observable) {
    return values[++state.index];
  } else {
    var _constructor2 = template && template.constructor;

    if (_constructor2 === Array) {
      var n = template.length;
      var next = Array(n);
      for (var i = 0; i < n; ++i) {
        next[i] = combine(template[i], values, state);
      }return next;
    } else if (_constructor2 === Object) {
      var _next = {};
      for (var k in template) {
        _next[k] = combine(template[k], values, state);
      }return _next;
    } else {
      return template;
    }
  }
}

function invoke(xs) {
  if (!(xs instanceof Array)) return xs;

  var nm1 = xs.length - 1;
  var f = xs[nm1];
  return f instanceof Function ? f.apply(undefined, _toConsumableArray(xs.slice(0, nm1))) : xs;
}

//

function Combine() {
  Kefir.Property.call(this);
}

Combine.prototype = Object.create(Kefir.Property.prototype);

Combine.prototype._maybeEmitValue = function (next) {
  var prev = this._currentEvent;
  if (!prev || !R.equals(prev.value, next)) this._emitValue(next);
};

//

function CombineMany(template, n) {
  Combine.call(this);
  this._template = template;
  this._handlers = n;
  this._values = null;
}

CombineMany.prototype = Object.create(Combine.prototype);

CombineMany.prototype._onActivation = function () {
  var template = this._template;
  var n = this._handlers;
  var handlers = Array(n);
  var values = Array(n);
  for (var i = 0; i < n; ++i) {
    values[i] = this;
    handlers[i] = this;
  }
  this._handlers = handlers;
  this._values = values;
  subscribe(template, handlers, this);
};

CombineMany.prototype._handleAny = function (i, e) {
  switch (e.type) {
    case "value":
      {
        var values = this._values;
        values[i] = e.value;
        for (var j = 0, n = values.length; j < n; ++j) {
          if (values[j] === this) return;
        }this._maybeEmitValue(invoke(combine(this._template, values, { index: -1 })));
        break;
      }
    case "error":
      {
        this._emitError(e.value);
        break;
      }
    default:
      {
        var handlers = this._handlers;
        handlers[i] = null;
        for (var _j = 0, _n = handlers.length; _j < _n; ++_j) {
          if (handlers[_j]) return;
        }this._handlers = handlers.length;
        this._values = null;
        this._emitEnd();
        break;
      }
  }
};

CombineMany.prototype._onDeactivation = function () {
  var handlers = this._handlers;
  this._handlers = handlers.length;
  this._values = null;
  unsubscribe(this._template, handlers);
};

//

function CombineOne(template) {
  Combine.call(this);
  this._template = template;
  this._handler = null;
}

CombineOne.prototype = Object.create(Combine.prototype);

CombineOne.prototype._onActivation = function () {
  var _this = this;

  var handler = function handler(e) {
    return _this._handleAny(e);
  };
  this._handler = handler;
  forEach(this._template, function (observable) {
    return observable.onAny(handler);
  });
};

CombineOne.prototype._handleAny = function (e) {
  switch (e.type) {
    case "value":
      this._maybeEmitValue(invoke(combine(this._template, [e.value], { index: -1 })));
      break;
    case "error":
      this._emitError(e.value);
      break;
    default:
      this._handler = null;
      this._emitEnd();
      break;
  }
};

CombineOne.prototype._onDeactivation = function () {
  var _handler = this._handler;

  this._handler = null;
  forEach(this._template, function (observable) {
    return observable.offAny(_handler);
  });
};

//

function CombineOneWith(observable, fn) {
  Combine.call(this);
  this._observable = observable;
  this._fn = fn;
  this._handler = null;
}

CombineOneWith.prototype = Object.create(Combine.prototype);

CombineOneWith.prototype._onActivation = function () {
  var _this2 = this;

  var handler = function handler(e) {
    return _this2._handleAny(e);
  };
  this._handler = handler;
  this._observable.onAny(handler);
};

CombineOneWith.prototype._handleAny = function (e) {
  switch (e.type) {
    case "value":
      this._maybeEmitValue(this._fn(e.value));
      break;
    case "error":
      this._emitError(e.value);
      break;
    default:
      this._handler = null;
      this._emitEnd();
      break;
  }
};

CombineOneWith.prototype._onDeactivation = function () {
  var _handler = this._handler;
  var _observable = this._observable;

  this._handler = null;
  _observable.offAny(_handler);
};

//

var asProperty = exports.asProperty = function asProperty() {
  for (var _len = arguments.length, template = Array(_len), _key = 0; _key < _len; _key++) {
    template[_key] = arguments[_key];
  }

  var n = count(template);
  switch (n) {
    case 0:
      return invoke(template);
    case 1:
      return template.length === 2 && template[0] instanceof Kefir.Observable && template[1] instanceof Function ? new CombineOneWith(template[0], template[1]) : new CombineOne(template);
    default:
      return new CombineMany(template, n);
  }
};
